internal/quickfort/dig.lua

```diff
@@ -1,26 +1,569 @@
 -- dig-related logic for the quickfort script
 --@ module = true
+--[[
+This file designates tiles with the same rules as the in-game UI. For example,
+if the tile is hidden, we designate blindly to avoid spoilers. If it's visible,
+the shape and material of the target tile affects whether the designation has
+any effect.
+]]
 
 if not dfhack_flags.module then
     qerror('this script cannot be called directly')
 end
 
+local utils = require('utils')
 local quickfort_common = reqscript('internal/quickfort/common')
 local log = quickfort_common.log
 
-function do_run(zlevel, grid)
-    local stats = nil
-    print('"quickfort run" not yet implemented for mode: dig')
+local function is_construction(tileattrs)
+    return tileattrs.material == df.tiletype_material.CONSTRUCTION
+end
+
+local function is_floor(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.FLOOR
+end
+
+local function is_diggable_floor(tileattrs)
+    return is_floor(tileattrs) or
+            tileattrs.shape == df.tiletype_shape.BOULDER or
+            tileattrs.shape == df.tiletype_shape.PEBBLES
+end
+
+local function is_wall(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.WALL
+end
+
+local function is_fortification(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.FORTIFICATION
+end
+
+local function is_up_stair(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.STAIR_UP
+end
+
+local function is_down_stair(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.STAIR_DOWN
+end
+
+local function is_removable_shape(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.RAMP or
+            tileattrs.shape == df.tiletype_shape.STAIR_UP or
+            tileattrs.shape == df.tiletype_shape.STAIR_UPDOWN
+end
+
+local function is_gatherable(tileattrs)
+    return tileattrs.shape == df.tiletype_shape.SHRUB
+end
+
+local hard_natural_materials = utils.invert({
+    df.tiletype_material.STONE,
+    df.tiletype_material.FEATURE,
+    df.tiletype_material.LAVA_STONE,
+    df.tiletype_material.MINERAL,
+    df.tiletype_material.FROZEN_LIQUID,
+})
+
+local function is_hard(tileattrs)
+    return hard_natural_materials[tileattrs.material]
+end
+
+local function is_smooth(tileattrs)
+    return tileattrs.special == df.tiletype_special.SMOOTH
+end
+
+local function get_engraving(pos)
+    -- scan through engravings until we find the one at this pos
```

Comment:
I double-checked the engravings module; turns out that it's very old, not exposed to Lua, and also doesn't have something to do this. The buildings module does something similar with `findAtTile()`, but it's faster (since it's written in C++) and caches its results. You could try caching here, although I suspect most forts won't have *too* many engravings.

Reply:
agree -- I'll just add a comment that it could be made faster with caching and see if anybody complains about the slowness